### ISSUE TITLE
Option to fetch customer Address from customer module even after submitting, if empty.

### ISSUE THREAD

**Issue by GopalKedia2020** [resilient-tech/india-compliance#1320]
### Is your feature request related to a problem? Please describe.

We have newly setup ERPNext, where we imported customers details from Excel using import tool, this didnt generated the address for these customer rather created primary address in Customer module itself. Now, while generating sales invoice, the customer address field remained empty and we are not able to update it now since invoice is already submitted. As per ERPNext support, allowing this using "Property Setter Doctype" will lead to violation of product warranty. Right now while submiting E-Invoice i am getting below error. 

### App Versions
```
{
	"email_delivery_service": "0.0.1",
	"erpnext": "14.49.0",
	"frappe": "14.56.0",
	"india_compliance": "14.19.4",
	"twilio_integration": "0.0.1"
}
```
### Route
```
Form/Sales Invoice/HP/S/4010/23-24
```
### Traceback
```
Traceback (most recent call last):
  File "apps/frappe/frappe/app.py", line 95, in application
    response = frappe.api.handle()
  File "apps/frappe/frappe/api.py", line 55, in handle
    return frappe.handler.handle()
  File "apps/frappe/frappe/handler.py", line 47, in handle
    data = execute_cmd(cmd)
  File "apps/frappe/frappe/handler.py", line 85, in execute_cmd
    return frappe.call(method, **frappe.form_dict)
  File "apps/frappe/frappe/__init__.py", line 1622, in call
    return fn(*args, **newargs)
  File "apps/india_compliance/india_compliance/gst_india/utils/e_invoice.py", line 206, in generate_e_invoice
    raise e
  File "apps/india_compliance/india_compliance/gst_india/utils/e_invoice.py", line 123, in generate_e_invoice
    data = EInvoiceData(doc).get_data()
  File "apps/india_compliance/india_compliance/gst_india/utils/e_invoice.py", line 468, in get_data
    self.set_party_address_details()
  File "apps/india_compliance/india_compliance/gst_india/utils/e_invoice.py", line 610, in set_party_address_details
    self.billing_address = self.get_address_details(
  File "apps/india_compliance/india_compliance/gst_india/utils/transaction_data.py", line 398, in get_address_details
    if address.gst_state_number == "97":  # For Other Territory
AttributeError: 'NoneType' object has no attribute 'gst_state_number'

```
### Request Data
```
{
	"type": "POST",
	"args": {
		"docname": "HP/S/4010/23-24",
		"force": true
	},
	"headers": {},
	"error_handlers": {},
	"url": "/api/method/india_compliance.gst_india.utils.e_invoice.generate_e_invoice"
}
```
### Response Data
```
{
	"exc_type": "DoesNotExistError",
	"exception": "AttributeError: 'NoneType' object has no attribute 'gst_state_number'"
}
```

### Describe the solution you'd like

Allowing adding Customer Address, Address, and Place of Supply after submit, if these fileds are empty from customer and address modules, so that invoice can get the details required and we can submit E-Invoices even if we miss to add address. 

### Describe alternatives you've considered

The immediate solution can be to cancel and amend the invoice but out invoice series is having 15 character and we are not able to amend it since that add 2 more alphabats. I tried using custom scripts to add address on reload but it didnt worked.

**Comment by vorasmit** [comment]
On submitting the invoice, Allowing an update of the address is impossible.

The address is a part of the database and is crucial from many perspectives even in taxation. This could even lead to changes in tax templates.

I am afraid the only solution I can see currently is to recreate the invoices from scratch if you wish to maintain the same naming.

**Comment by GopalKedia2020** [comment]
![Screenshot 27-11-2023 at 04 02 pm](https://github.com/resilient-tech/india-compliance/assets/136687952/b2454cd2-c468-4179-82df-a7e0004c0c00)
I think, my message was not clear. As you can see in the screenshot, Place of supply, GST Numbers are already fetched. I don't want to change that. But the address part is missing, because the address is not set to mandatory in ERPNext while creating a new invoice and now, its not allowing me to upload the E-Invoice because of the error as shown in screenshot.

So, if you allow me to update the address here from address module for the same customer, whose GST number already fetched, my issue will be resolved. We just have few days left to submit all E-Invoices.

**Comment by vorasmit** [comment]
Hi,

Let me guess,
Is it possible that the first invoice was created, and then the India Compliance App was installed?

We do have validation to ensure that an address is required if an e-invoice is supposed to be created.

You will need to use a developer's help externally to help you with your particular case. We cannot allow updating addresses from the India Compliance App, as it would then affect all users.

**Comment by GopalKedia2020** [comment]
India Compliance was installed from first day. But Enable e-Invoicing was not checked on GST India Settings page.

As a workaround, we are now uploading the failed invoices manually using json file on GST portal. There is one request, if it is possible, can we fetch the IRN number from the portal for those failed invoices too, as we will be submitting those manually? Right now, it's still showing an error, even after one invoice is already submitted manually for testing.

Also, as you said, you have a system to validate it on submit if E-Invoice is checked, I think there is an issue with that too, which I thought will create a new git request but telling here only. These are several invoices which describe it as not applicable but those are eligible for E-Invoice, and I do get option to upload E-Inovice for those too in invoice detail page but in e-Invoice Status it says Not applicable. These invoices may not trigger the validation (need testing) but in filter there is definitely an issue.

**Comment by vorasmit** [comment]
We appreciate your feedback.

Certain fields are mandatory only for e-Invoice and we do validate it conditionally if e-Invoice is enabled (like customer address)

Kindly create a new issue with sample invoice details where you are seeing it as Not Applicable for newly generated invoices after you have enabled e-Invoice.

Note: e-Invoice applicability (status field) is a field in Sales Invoice. This is evaluated only once when you submit the invoice. If at the time of submission of the invoice, if e-Invoice setting was disabled, you should see it as Not Applicable. However, the e-invoice Summary report is dynamic and should show you all invoices where the e-invoice is applicable as of the date.

**Comment by GopalKedia2020** [comment]
While checking for the issue after receiving your message, I didn't find any error in "e-Invoice Status" filter after I turned on the E-Invoice setting. So, it seems, there is no issues with filters once all the settings are set correctly in ERPNext. 

I hope, you have considered my request to fetch the IRN no for all failed E-Invoice uploads. This can be achieved, if we first check for the Invoice number from the GST portal (if status is failed), and if the invoice number is present on GST Portal, then it can fetch the IRN no and mark it success. If no data found for that invoice, it can trigger the upload functions and show errors / success as its working now.

**Comment by vorasmit** [comment]
- We do have fallbacks and we do fetch IRN if it's manually generated.
- However, it's still part of the Generate IRN feature which is failing in your case.

The best solution is to update your records in ERPNext with the correct address.

If you need developer assistance for this and are not in touch with any, reach out to us info@resilient.tech and we'll help you out.
